### PR TITLE
emit_x64_vector: AVX512 implementation of EmitVectorNot

### PR DESCRIPTION
**Pull request by Wunkolo**
Single in-place ternary logic instruction to invert bits of a vector-register.

~~I hope all these AVX512 PRs don't make me look like I have stock in Intel or something~~

### CODE REVIEW

**Comment by MerryMage** [comment]
I'm kinda on the border with this one. (The original implementation already averages 1 cycle.)

**Comment by lioncash** [comment]
Same

**Comment by Wunkolo** [comment]
Here's some Skylake-X data:
Instruction | Lat | TP | Uops | Ports
-|-|-|-|-
[VPTERNLOGQ (XMM, XMM, XMM, I8)](https://uops.info/html-instr/VPTERNLOGQ_XMM_XMM_XMM_I8.html) | [1](https://uops.info/html-lat/SKX/VPTERNLOGQ_XMM_XMM_XMM_I8-Measurements.html) | 0.33/ [0.33](https://uops.info/html-tp/SKX/VPTERNLOGQ_XMM_XMM_XMM_I8-Measurements.html) | [1 / 1](https://uops.info/html-tp/SKX/VPTERNLOGQ_XMM_K_XMM_XMM_I8-Measurements.html) | [1*p015](https://uops.info/html-ports/SKX/VPTERNLOGQ_XMM_XMM_XMM_I8-Measurements.html)

Instruction | Lat | TP | Uops | Ports
-|-|-|-|-
[PCMPEQW (XMM, XMM)](https://uops.info/html-instr/PCMPEQW_XMM_XMM.html) | [[0;1]](https://uops.info/html-lat/SKX/PCMPEQW_XMM_XMM-Measurements.html) | 0.50 / [0.50](https://uops.info/html-tp/SKX/PCMPEQW_XMM_XMM-Measurements.html) | [1 / 1](https://uops.info/html-tp/SKX/PCMPEQW_XMM_XMM-Measurements.html) | [1*p01](https://uops.info/html-ports/SKX/PCMPEQW_XMM_XMM-Measurements.html)
[PXOR (XMM, XMM)](https://uops.info/html-instr/PXOR_XMM_XMM.html) | [[0;1]](https://uops.info/html-lat/SKX/PXOR_XMM_XMM-Measurements.html) | 0.33 / [0.33](https://uops.info/html-tp/SKX/PXOR_XMM_XMM-Measurements.html) | [1 / 1](https://uops.info/html-tp/SKX/PXOR_XMM_XMM-Measurements.html) | [1*p015](https://uops.info/html-ports/SKX/PXOR_XMM_XMM-Measurements.html)

[Nanobench](https://github.com/andreas-abel/nanoBench) I took on my [i9-7900x](https://en.wikichip.org/wiki/intel/core_i9/i9-7900x):

```
% sudo ./kernel-nanoBench.sh -unroll 500 -warm_up_count 10 -asm "pcmpeqw xmm1, xmm1; pxor xmm0, xmm1" -config configs/cfg_Skylake_common.txt
RDTSC: 0.92
Instructions retired: 2.00
Core cycles: 1.00
Reference cycles: 0.92
UOPS_ISSUED.ANY: 2.00
UOPS_EXECUTED.THREAD: 2.00
UOPS_DISPATCHED_PORT.PORT_0: 0.72
UOPS_DISPATCHED_PORT.PORT_1: 0.00
UOPS_DISPATCHED_PORT.PORT_2: 0.00
UOPS_DISPATCHED_PORT.PORT_3: 0.00
UOPS_DISPATCHED_PORT.PORT_4: 0.00
UOPS_DISPATCHED_PORT.PORT_5: 0.00
UOPS_DISPATCHED_PORT.PORT_6: 0.00
UOPS_DISPATCHED_PORT.PORT_7: 0.00
BR_INST_RETIRED.ALL_BRANCHES: 0.00
BR_MISP_RETIRED.ALL_BRANCHES: 0.00
MEM_LOAD_RETIRED.L1_HIT: 0.00
MEM_LOAD_RETIRED.L1_MISS: 0.00
MEM_LOAD_RETIRED.L2_HIT: 0.00
MEM_LOAD_RETIRED.L2_MISS: 0.00
MEM_LOAD_RETIRED.L3_HIT: 0.00
MEM_LOAD_RETIRED.L3_MISS: 0.00
```

```
% sudo ./kernel-nanoBench.sh -unroll 500 -warm_up_count 10 -asm "vpternlogd xmm0, xmm0, xmm0, 0x55" -config configs/cfg_Skylake_common.txt
RDTSC: 0.81
Instructions retired: 1.00
Core cycles: 1.00
Reference cycles: 0.83
UOPS_ISSUED.ANY: 1.00
UOPS_EXECUTED.THREAD: 1.00
UOPS_DISPATCHED_PORT.PORT_0: 0.33
UOPS_DISPATCHED_PORT.PORT_1: 0.00
UOPS_DISPATCHED_PORT.PORT_2: 0.00
UOPS_DISPATCHED_PORT.PORT_3: 0.00
UOPS_DISPATCHED_PORT.PORT_4: 0.00
UOPS_DISPATCHED_PORT.PORT_5: 0.00
UOPS_DISPATCHED_PORT.PORT_6: 0.00
UOPS_DISPATCHED_PORT.PORT_7: 0.00
BR_INST_RETIRED.ALL_BRANCHES: 0.00
BR_MISP_RETIRED.ALL_BRANCHES: 0.00
MEM_LOAD_RETIRED.L1_HIT: 0.00
MEM_LOAD_RETIRED.L1_MISS: 0.00
MEM_LOAD_RETIRED.L2_HIT: 0.00
MEM_LOAD_RETIRED.L2_MISS: 0.00
MEM_LOAD_RETIRED.L3_HIT: 0.00
MEM_LOAD_RETIRED.L3_MISS: 0.00
```

Primarily there are uop, throughput, and port-pressure gains which might benefit the context that the instruction is emitted in I guess 🥴

Your call

File: src/dynarmic/backend/x64/emit_x64_vector.cpp [hunk 1]
```diff
@@ -2137,12 +2137,17 @@ void EmitX64::EmitVectorNarrow64(EmitContext& ctx, IR::Inst* inst) {
 void EmitX64::EmitVectorNot(EmitContext& ctx, IR::Inst* inst) {
     auto args = ctx.reg_alloc.GetArgumentInfo(inst);
 
+    if (code.HasHostFeature(HostFeature::AVX512_Ortho)) {
+        const Xbyak::Xmm xmm_a = ctx.reg_alloc.UseScratchXmm(args[0]);
+        code.vpternlogq(xmm_a, xmm_a, xmm_a, u8(~Tern::c));
```
Comment (MerryMage) [inline src/dynarmic/backend/x64/emit_x64_vector.cpp:2142]:
```cpp
        const Xbyak::Xmm result = ctx.reg_alloc.ScratchXmm();
        const Xbyak::Xmm operand = ctx.reg_alloc.UseXmm(args[0]);
        code.vpternlogq(result, operand, operand, u8(~Tern::c));
```
Avoid unnecessary moves in the register allocator if not required.

**Pull request merged**